diff --git a/common.gypi b/common.gypi
index 47c1d0c..c7791b2 100644
--- a/common.gypi
+++ b/common.gypi
@@ -250,7 +250,7 @@
           },],
           ['OS=="win"', {
             'conditions': [
-              ['enable_lto=="true"', {
+              ['configuring_node==1 and enable_lto=="true"', {
                 'msvs_settings': {
                   'VCCLCompilerTool': {
                     'AdditionalOptions': ['-flto=full'],
@@ -263,7 +263,7 @@
                   },
                 },
               },],
-              ['enable_thin_lto=="true"', {
+              ['configuring_node==1 and enable_thin_lto=="true"', {
                 'msvs_settings': {
                   'VCCLCompilerTool': {
                     'AdditionalOptions': ['-flto=thin'],
@@ -276,7 +276,7 @@
                   },
                 },
               },],
-              ['(enable_thin_lto=="true" or enable_lto=="true") and lto_jobs!=""', {
+              ['configuring_node==1 and (enable_thin_lto=="true" or enable_lto=="true") and lto_jobs!=""', {
                 'msvs_settings': {
                   'VCLinkerTool': {
                     'AdditionalOptions': ['/opt:lldltojobs=<(lto_jobs)'],
